File: src/java/main/java/com/pigishentertainment/dndassistant/domain/Monster.java
package com.pigishentertainment.dndassistant.domain;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.time.Instant;

/**
 * Mirrors the bundled JSON shape in
 * {@code src/res/resources/monster_manual_monsters.json} (and the
 * frontend {@code Monster} type). Field names use PascalCase on the wire
 * (AC, HP, Speed, INT, ...) and snake_case for the multi-word fields
 * (Saving_Throws, Legendary_Actions, ...). DB columns are lowercase with
 * trailing underscores; the repository maps between them.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Monster {
  private Long id;

  @JsonProperty("name")    private String name;
  @JsonProperty("meta")    private String meta;
  @JsonProperty("AC")      private String ac;
  @JsonProperty("HP")      private String hp;
  @JsonProperty("Speed")   private String speed;

  @JsonProperty("STR")       private String str;
  @JsonProperty("STR_mod")   private String str_mod;
  @JsonProperty("DEX")       private String dex;
  @JsonProperty("DEX_mod")   private String dex_mod;
  @JsonProperty("CON")       private String con;
  @JsonProperty("CON_mod")   private String con_mod;
  @JsonProperty("INT")       private String intValue;
  @JsonProperty("INT_mod")   private String int_mod;
  @JsonProperty("WIS")       private String wis;
  @JsonProperty("WIS_mod")   private String wis_mod;
  @JsonProperty("CHA")       private String cha;
  @JsonProperty("CHA_mod")   private String cha_mod;

  @JsonProperty("Saving_Throws")         private String saving_throws;
  @JsonProperty("Skills")                private String skills;
  @JsonProperty("Damage_Vulnerabilities") private String damage_vulnerabilities;
  @JsonProperty("Damage_Resistances")    private String damage_resistances;
  @JsonProperty("Damage_Immunities")     private String damage_immunities;
  @JsonProperty("Condition_Immunities")  private String condition_immunities;
  @JsonProperty("Senses")                private String senses;
  @JsonProperty("Languages")             private String languages;
  @JsonProperty("Challenge")             private String challenge;

  @JsonProperty("Traits")            private String traits;
  @JsonProperty("Actions")           private String actions;
  @JsonProperty("Reactions")         private String reactions;
  @JsonProperty("Legendary_Actions") private String legendary_actions;
  @JsonProperty("description")       private String description;
  @JsonProperty("Lair_Actions")      private String lair_actions;
  @JsonProperty("Regional_Effects")  private String regional_effects;
  @JsonProperty("img_url")           private String img_url;

  @JsonProperty("provenance")    private String provenance = "homebrew";
  @JsonProperty("owner_user_id") private String owner_user_id;
  @JsonProperty("created_at")    private Instant created_at;
  @JsonProperty("updated_at")    private Instant updated_at;

  public Long getId() { return id; }
  public void setId(Long id) { this.id = id; }

  public String getName() { return name; }
  public void setName(String name) { this.name = name; }

  public String getMeta() { return meta; }
  public void setMeta(String meta) { this.meta = meta; }

  public String getAc() { return ac; }
  public void setAc(String ac) { this.ac = ac; }

  public String getHp() { return hp; }
  public void setHp(String hp) { this.hp = hp; }

  public String getSpeed() { return speed; }
  public void setSpeed(String speed) { this.speed = speed; }

  public String getStr() { return str; }
  public void setStr(String str) { this.str = str; }
  public String getStr_mod() { return str_mod; }
  public void setStr_mod(String str_mod) { this.str_mod = str_mod; }

  public String getDex() { return dex; }
  public void setDex(String dex) { this.dex = dex; }
  public String getDex_mod() { return dex_mod; }
  public void setDex_mod(String dex_mod) { this.dex_mod = dex_mod; }

  public String getCon() { return con; }
  public void setCon(String con) { this.con = con; }
  public String getCon_mod() { return con_mod; }
  public void setCon_mod(String con_mod) { this.con_mod = con_mod; }

  public String getInt_() { return intValue; }
  public void setInt_(String intValue) { this.intValue = intValue; }
  public String getInt_mod() { return int_mod; }
  public void setInt_mod(String int_mod) { this.int_mod = int_mod; }

  public String getWis() { return wis; }
  public void setWis(String wis) { this.wis = wis; }
  public String getWis_mod() { return wis_mod; }
  public void setWis_mod(String wis_mod) { this.wis_mod = wis_mod; }

  public String getCha() { return cha; }
  public void setCha(String cha) { this.cha = cha; }
  public String getCha_mod() { return cha_mod; }
  public void setCha_mod(String cha_mod) { this.cha_mod = cha_mod; }

  public String getSaving_throws() { return saving_throws; }
  public void setSaving_throws(String saving_throws) { this.saving_throws = saving_throws; }
  public String getSkills() { return skills; }
  public void setSkills(String skills) { this.skills = skills; }
  public String getDamage_vulnerabilities() { return damage_vulnerabilities; }
  public void setDamage_vulnerabilities(String v) { this.damage_vulnerabilities = v; }
  public String getDamage_resistances() { return damage_resistances; }
  public void setDamage_resistances(String v) { this.damage_resistances = v; }
  public String getDamage_immunities() { return damage_immunities; }
  public void setDamage_immunities(String v) { this.damage_immunities = v; }
  public String getCondition_immunities() { return condition_immunities; }
  public void setCondition_immunities(String v) { this.condition_immunities = v; }
  public String getSenses() { return senses; }
  public void setSenses(String senses) { this.senses = senses; }
  public String getLanguages() { return languages; }
  public void setLanguages(String languages) { this.languages = languages; }
  public String getChallenge() { return challenge; }
  public void setChallenge(String challenge) { this.challenge = challenge; }

  public String getTraits() { return traits; }
  public void setTraits(String traits) { this.traits = traits; }
  public String getActions() { return actions; }
  public void setActions(String actions) { this.actions = actions; }
  public String getReactions() { return reactions; }
  public void setReactions(String reactions) { this.reactions = reactions; }
  public String getLegendary_actions() { return legendary_actions; }
  public void setLegendary_actions(String v) { this.legendary_actions = v; }
  public String getDescription() { return description; }
  public void setDescription(String description) { this.description = description; }
  public String getLair_actions() { return lair_actions; }
  public void setLair_actions(String v) { this.lair_actions = v; }
  public String getRegional_effects() { return regional_effects; }
  public void setRegional_effects(String v) { this.regional_effects = v; }
  public String getImg_url() { return img_url; }
  public void setImg_url(String img_url) { this.img_url = img_url; }

  public String getProvenance() { return provenance; }
  public void setProvenance(String provenance) { this.provenance = provenance; }

  public String getOwner_user_id() { return owner_user_id; }
  public void setOwner_user_id(String owner_user_id) { this.owner_user_id = owner_user_id; }

  public Instant getCreated_at() { return created_at; }
  public void setCreated_at(Instant created_at) { this.created_at = created_at; }
  public Instant getUpdated_at() { return updated_at; }
  public void setUpdated_at(Instant updated_at) { this.updated_at = updated_at; }
}

File: src/ts/types/Monster.ts
// Monsters are loaded from the backend API (see src/ts/api/monsters.ts).
// The bundled SRD / Monster Manual files are still used as seed sources
// by `worker/seed/generate.mjs`, but the frontend reads from
// `/api/v1/monsters` so any user-created (homebrew) monsters show up
// too.
//
// Wire shape (snake_case) matches the Worker route at
// `worker/src/routes/monsters.ts` and the D1 column names. This is the
// same convention as `Spell` and `Gear` — the previous PascalCase
// type matched the bundled JSON, but since the API is the source of
// truth, the type mirrors the API.

export const PLACEHOLDER_IMAGES = {
  humanoid: "https://media-waterdeep.cursecdn.com/attachments/2/656/humanoid.jpg",
  undead: "https://media-waterdeep.cursecdn.com/attachments/2/660/undead.jpg",
  beast: "https://media-waterdeep.cursecdn.com/attachments/2/648/beast.jpg",
  plant: "https://media-waterdeep.cursecdn.com/attachments/2/659/plant.jpg",
  ooze: "https://media-waterdeep.cursecdn.com/attachments/2/658/ooze.jpg"
};

export const resolveImgUrl = (imgUrl: string): string => {
  if (!imgUrl) return PLACEHOLDER_IMAGES.humanoid;
  if (/^https?:\/\//i.test(imgUrl)) return imgUrl;
  return `${process.env.PUBLIC_URL || ''}/${imgUrl}`;
};

export type Monster = {
  id?: number,
  name: string,
  meta: string,
  ac: string,
  hp: string,
  speed: string,
  str: string,
  str_mod: string,
  dex: string,
  dex_mod: string,
  con: string,
  con_mod: string,
  int: string,
  int_mod: string,
  wis: string,
  wis_mod: string,
  cha: string,
  cha_mod: string,
  saving_throws: string | undefined,
  skills: string | undefined,
  damage_vulnerabilities: string | undefined,
  damage_resistances: string | undefined,
  damage_immunities: string | undefined,
  condition_immunities: string | undefined,
  senses: string,
  languages: string,
  challenge: string,
  traits: string | undefined,
  actions: string,
  reactions: string | undefined,
  legendary_actions: string | undefined,
  description?: string,
  lair_actions?: string,
  regional_effects?: string,
  img_url: string,
  provenance: string,
};

export const defaultMonster = {
  name: "",
  meta: "Medium humanoid, any",
  ac: "12",
  hp: "10",
  speed: "30ft.",
  str: "10",
  str_mod: "(+0)",
  dex: "10",
  dex_mod: "(+0)",
  con: "10",
  con_mod: "(+0)",
  int: "10",
  int_mod: "(+0)",
  wis: "10",
  wis_mod: "(+0)",
  cha: "10",
  cha_mod: "(+0)",
  senses: "Passive Perception 10",
  languages: "--",
  challenge: "1/4 (50 XP)",
  actions: "",
  img_url: PLACEHOLDER_IMAGES.humanoid,
  provenance: "homebrew",
};
